Convert inline documentation to JSDoc

fredrikekelund opened the pull request:
### What this PR changes
It introduces JSDoc documentation format for documenting methods, events and class members. This is still a **work in progress** (only the main crawler module is done), but given the magnitude of the changes here, I wanted to open a PR so we can iron out any issues as early as possible.

I've tried generating a documentation site from this and it actually turned out quite nicely!

### Rationale
See #258

fredrikekelund commented:
I pushed some documentation for the queue module. I'll try and push some for the cookie module tomorrow as well, but after that I think we should try and add additional documentation incrementally. What we have at this point will help both ordinary developers who want to use simplecrawler as well as those who would like to write third party queue implementations

fredrikekelund commented:
Alright, that's the cookie module documented as well! The diff here is pretty big by now, and I've snuck some changes in functionality in there as well. I'll try to leave a line comment on all of those to highlight them if you want to have a look, @cgiffard

fredrikekelund commented on lib/cookies.js:
```diff
@@ -1,63 +1,49 @@
+     * @private
+     * @type {Array}
+     */
+    this.cookies = [];
```
Was there a particular reason for the custom getter setup with the cookies array previously? Otherwise, this format worked right off the bat and makes it a bit easier to document and understand

cgiffard replied:
I wish I could remember! I see no reason not to change it now. :)

fredrikekelund commented on lib/cookies.js:
```diff
@@ -70,130 +56,131 @@ CookieJar.prototype.add = function(name, value, expiry, path, domain, httponly,
-
         // If the names don't match, we're not removing this cookie
-        if (Boolean(name) && cookie.name !== name) {
+        if (typeof name === "string" && cookie.name !== name) {
```
Thought I'd express intent a bit clearer here

cgiffard replied:
FWIW I kinda read this line as "Ah, we're testing it to see if it's a string, so it's probably a polymorphic variable." That said though, it's pretty clearly documented, so I don't see any problem with continuing as is. :)

fredrikekelund replied:
Think you're right on this one, reverted it!

fredrikekelund commented on lib/crawler.js:
```diff
@@ -50,124 +40,276 @@ var Crawler = function(initialURL) {
+     * Controls which URL to request first
+     * @type {String}
+     */
+    this.initialURL = initialURL;
```
Assigning properties directly to `this` rather than to a reference to `this` (ie. `crawler`) makes it so that JSDoc can understand what's a class member

fredrikekelund commented on lib/crawler.js:
```diff
@@ -297,39 +465,34 @@ Crawler.prototype.start = function() {
+ * @return {Boolean}    Returns true if the URL is allowed to be fetched, otherwise false
  */
-Crawler.prototype.urlIsAllowed = function (queueItem) {
+Crawler.prototype.urlIsAllowed = function(url) {
```
I think this was my attempt to make this method more flexible at some point. Coming back to it now, though, I just found it to make the method more confusing. And when calling this method, you're always going to have access to the queue item's URL if you have access to the queue item itself

fredrikekelund commented on lib/crawler.js:
```diff
@@ -1380,21 +1474,27 @@ Crawler.prototype.handleResponse = function(queueItem, response, timeCommenced)
+        redirectQueueItem = crawler.processURL(response.headers.location, queueItem);
 
-        // Emit redirect event
-        crawler.emit("fetchredirect", queueItem, referrerQueueItem, response);
```
This used to be plain wrong. `referrerQueueItem` was always the queue item for the next request, so I fixed the wording here

fredrikekelund commented on lib/crawler.js:
```diff
@@ -1096,17 +1154,19 @@ Crawler.prototype.fetchQueueItem = function(queueItem) {
-        // but we don't want to do anything in that case.
-        if (clientRequest._crawlerHandled) {
+    clientRequest.on("error", function(error) {
+        if (clientRequest.aborted) {
```
We can rely on this property set by node itself when a request is aborted rather than a manual one. It expresses intent more clearly as well

fredrikekelund commented on lib/crawler.js:
```diff
@@ -224,28 +383,44 @@ var Crawler = function(initialURL) {
-        _fetchConditions: [],
-        _openListeners: 0
+        _robotsTxts: [],
+        _touchedHosts: []
```
I made `touchedHosts` a "private" member, since it really is meant only for internal use by the crawler

fredrikekelund commented on lib/queue.js:
```diff
@@ -41,21 +30,79 @@ function compare (a, b) {
+ * @class
+ */
 var FetchQueue = function() {
-    this.oldestUnfetchedIndex = 0;
```
I made `referrerQueueItem`, `completeCache` and `scanIndex` non-enumerable properties and renamed them to start with an underscore. We don't use them when defrosting a queue, so it makes sense to not include them in the frozen JSON file either (and this is a good way of avoiding that)

fredrikekelund commented on lib/queue.js:
```diff
@@ -41,21 +30,79 @@ function compare (a, b) {
-module.exports = FetchQueue;
+FetchQueue.prototype = Array.prototype;
 
-FetchQueue.prototype = [];
```
Previously, if for some reasons multiple queue instances would have been created, they would've inherited from the same array. Now they inherit from `Array.prototype` instead

fredrikekelund commented on lib/queue.js:
```diff
@@ -41,21 +30,79 @@ function compare (a, b) {
+ * @param {FetchQueue~addCallback} callback
+ */
 FetchQueue.prototype.add = function(queueItem, force, callback) {
-    callback = callback instanceof Function ? callback : noop;
```
In the current version of the queue, callbacks aren't optional, so I've removed all of these lines

fredrikekelund commented on package.json:
```diff
@@ -28,7 +28,8 @@
     "mocha": "mocha -R spec -t 5000",
-    "test": "npm run lint && npm run mocha"
+    "test": "npm run lint && npm run mocha",
+    "docs": "jsdoc -c jsdoc.json"
```
Give it a try! :)

cgiffard commented:
In the interests of getting a release out, I'd be keen to pull this off the 1.0 release and push it to 1.1. :)

cgiffard commented:
(Doing that now, unless there are strong arguments to the contrary!)

(There's no reason we can't get the 1.1 out quickly afterward.)

cgiffard commented:
Actually, 1.0 looks pretty imminent. Let's just get this over and done with!

fredrikekelund commented:
Started working on fixing the merge conflicts here yesterday evening, but they were pretty major, so I didn't have the time to finish. Will push the fixes later today!

fredrikekelund commented:
Alright, I fixed the conflicts and fixed some stuff I had previously missed. Think we're good to merge this, @cgiffard?

FYI, I used the DocBlockr text editor plugin (I think it's available for several editors - at least Sublime and Atom) to help easily generate JSDoc templates for methods and members. If it turns out that it actually is subjective in how it formats comments in some way (although I don't think that's the case), then we know where that came from

cgiffard commented:
I like it. I think we can merge. There are some style things I'd like to tweak a little — adding a bit more vertical whitespace in, and to try and limit line length a little, but they're not impediments for merging! From my perspective, press the button whenever you're ready to go! I'm excited. :)

Thank you again!
